File: jsSolutions/numOfIslands.js
/*
Problem description
Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.

An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.
problem link: https://leetcode.com/problems/number-of-islands/

*/


/**
 * @param {character[][]} grid
 * @return {number}
 */
 var numIslands = function(grid) {
    let row = grid.length;
    let col = grid[0].length;
    let islands = 0;
    for(let i=0;i<row;i++)
        {
            for(let j = 0;j<col;j++)
                {
                    if(grid[i][j] === '1')
                        {
                            islands++;
                            dfs(grid,i,j);
                        }
                    
                }
        }
    return islands;
    
};
var dfs = function(grid,i,j) {
    let row = grid.length;
    let col = grid[0].length;
    if(i>=0 && i<row && j>=0 && j<col && grid[i][j] === '1')
        {
            grid[i][j] ='0';
            dfs(grid,i+1,j); // up
            dfs(grid,i-1,j); //down
            dfs(grid,i,j+1);//right
            dfs(grid,i,j-1); //left
            
        }
    
} 

File: cppSolutions/numOfIslands.cpp

/*
Problem description
Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.

An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.
problem link: https://leetcode.com/problems/number-of-islands/

*/
#include<vector>;
using namespace std;
void dfs(vector<vector<char>>& grid,int row,int col)
{
    if(row>=0 && row<grid.size() && col>=0 && col<grid[row].size() && grid[row][col] == '1' )
    {
        grid[row][col] ='0';
        dfs(grid,row+1,col);
        dfs(grid,row-1,col);
        dfs(grid,row,col+1);
        dfs(grid,row,col-1);
    }
    else
    {
        return;
     }
}
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int rowNumber = grid.size();
        int colNumber = grid[0].size();
        int count =0;
        for(int i = 0;i<rowNumber;i++)
        {
            for(int j = 0;j<colNumber;j++)
            {
                if(grid[i][j]=='1')
                    count++;
                dfs(grid,i,j);
            }
        }
        return count;
        
    }
};